datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,344 @@ fn optimize_join(
+                        let expr = match &projection.expr[i] {
+                            Expr::Alias(expr, _) => expr.as_ref().clone(),
+                            expr => expr.clone(),
+                        };
+
+                        // Convert both qualified and unqualified fields
+                        [
+                            (field.name().clone(), expr.clone()),
+                            (field.qualified_name(), expr),
+                        ]
                     })
-                    .collect::<Result<Vec<_>>>()?;
-                state.filters.extend(join_side_filters);
+                    .collect::<HashMap<_, _>>();
+
+                // re-write all filters based on this projection
+                // E.g. in `Filter: b\n  Projection: a > 1 as b`, we can swap them, but the filter must be "a > 1"
+                let new_filter = LogicalPlan::Filter(Filter::try_new(
+                    replace_cols_by_name(filter.predicate().clone(), &replace_map)?,
+                    projection.input.clone(),
+                )?);
+
+                // optimize inner
+                from_plan(
+                    filter.input(),
+                    &(**filter.input()).expressions(),
+                    &[new_filter],
+                )?
             }
-
-            optimize_join(state, plan, left, right, on_filters)
-        }
-        LogicalPlan::TableScan(TableScan {
-            source,
-            projected_schema,
-            filters,
-            projection,
-            table_name,
-            fetch,
-        }) => {
-            let mut used_columns = HashSet::new();
-            let mut new_filters = filters.clone();
-
-            for (filter_expr, cols) in &state.filters {
-                let (preserve_filter_node, add_to_provider) =
-                    match source.supports_filter_pushdown(filter_expr)? {
-                        TableProviderFilterPushDown::Unsupported => (true, false),
-                        TableProviderFilterPushDown::Inexact => (true, true),
-                        TableProviderFilterPushDown::Exact => (false, true),
-                    };
-
-                if preserve_filter_node {
-                    used_columns.extend(cols.clone());
+            LogicalPlan::Union(union) => {
+                let mut inputs = Vec::with_capacity(union.inputs.len());
+                for input in &union.inputs {
+                    let replace_map = input
+                        .schema()
+                        .fields()
+                        .iter()
+                        .map(|field| (field.qualified_name(), col(field.name())))
+                        .collect::<HashMap<_, _>>();
+
+                    let push_predicate =
+                        replace_cols_by_name(filter.predicate().clone(), &replace_map)?;
+                    inputs.push(Arc::new(LogicalPlan::Filter(Filter::try_new(
+                        push_predicate,
```

Comment:
Looks like the `replace_map` for Union is not correct. My understanding for Union is we always use the left side field names as the output schema of Union. If we try to push down Filter to union's children,  need to replace the column in the Filter with the field name of each union's input.  The mapping should be based on the field index.

Reply:
My mistake, fixed it.